src/Connection.php

```diff
@@ -378,7 +378,7 @@ private function addCriteriaCondition(
      * @param array<string, mixed>                                                           $criteria
      * @param array<int, string|ParameterType|Type>|array<string, string|ParameterType|Type> $types
      *
-     * @return int|string The number of affected rows.
+     * @return non-negative-int|numeric-string The number of affected rows.
```

Comment:
We cannot guarantee that the integer won't be negative. In fact, some drivers might report errors by returning `-1`.

Reply:
IMO, we should.
Handling the errors / exceptions by our side and throwing early, I think we could prevent our countable methods to return negative or non integer values.
By instance, using checks like `$result === false`, as in this case:
https://github.com/doctrine/dbal/blob/210b7df961aa40b4790d77f00b031f38e803f106/src/Driver/Mysqli/Connection.php#L55-L68

Reply:
> IMO, we should.

No, we shouldn't.

> Handling the errors / exceptions by our side and throwing early,

That would mean that we would throw on `Connection::executeStatement()` if the query was executed successfully and the driver just failed to calculate the number of affected rows even if the caller does not care about that number. This would be really bad.

Reply:
No, sorry for the lack of context. I'm not proposing to block the execution of the statement, but adding the required checks to guarantee that methods like `Result::rowCount()` return non negative integer values. The first approach I can think of is performing the check in the same method, and throwing an exception if the value is not in that range.
IMO, is counter-intuitive for a method intended to count, to return negative values. I think that's similar to the old  PHP API where functions were returning values like `false`, `null` or `-1` to indicate the existence of an error, instead of using a proper exception mechanism. Given DBAL is intended to abstract the database API, I guess we could use an exception when we know an error occurred.

Reply:
Yes, I'll try to work on that in a different PR. I'm just performing some tests here until the PR can be reach a mergeable status.
Thank you for your concerns and explanations.